views/default/page/layouts/widgets.php

```diff
@@ -35,64 +35,82 @@
 
 $context = elgg_get_context();
 
-$widget_types = elgg_get_widget_types([
-	'context' => $context,
-	'exact' => $exact_match,
-	'container' => $owner,
-]);
-
-elgg_push_context('widgets');
-
 $widgets = elgg_get_widgets($owner->guid, $context);
 
-$layout_attrs = elgg_format_attributes(array(
-	'class' => 'elgg-layout-widgets',
-	'data-page-owner-guid' => $owner->guid,
-));
+$result = '';
+$no_widgets = elgg_extract('no_widgets', $vars);
+if (empty($widgets) && !empty($no_widgets)) {
+	if ($no_widgets instanceof \Closure) {
+		echo $no_widgets();
+	} else {
+		$result .= $no_widgets;
+	}
+}
+
+if ($show_add_widgets && elgg_can_edit_widget_layout($context)) {
+	$result .= elgg_view('page/layouts/widgets/add_button', $vars);
+}
 
-echo "<div $layout_attrs>";
+// push context after the add_button as add button uses current context
+elgg_push_context('widgets');
 
-if (elgg_can_edit_widget_layout($context)) {
-	if ($show_add_widgets) {
-		echo elgg_view('page/layouts/widgets/add_button');
-	}
-	$params = array(
-		'widgets' => $widgets,
+if ($widgets) {
+	$widget_types = elgg_get_widget_types([
 		'context' => $context,
-		'exact_match' => $exact_match,
-		'show_access' => $show_access,
-	);
-	echo elgg_view('page/layouts/widgets/add_panel', $params);
+		'exact' => $exact_match,
+		'container' => $owner,
+	]);
 }
 
-if (isset($vars['content'])) {
-	echo $vars['content'];
+$widget_class = "elgg-col-1of{$num_columns}";
+
+// move hidden columns widgets to last visible column
+if (!isset($widgets[$num_columns])) {
+	$widgets[$num_columns] = [];
 }
 
-$widget_class = "elgg-col-1of{$num_columns}";
-for ($column_index = 1; $column_index <= $num_columns; $column_index++) {
-	if (isset($widgets[$column_index])) {
-		$column_widgets = $widgets[$column_index];
-	} else {
-		$column_widgets = array();
+foreach ($widgets as $index => $column_widgets) {
+	if ($index <= $num_columns) {
+		continue;
+	}
+		
+	// append widgets to last column and retain order
+	foreach ($column_widgets as $column_widget) {
+		$widgets[$num_columns][] = $column_widget;
 	}
+	unset($widgets[$index]);
+}
 
-	echo "<div class=\"$widget_class elgg-widgets\" id=\"elgg-widget-col-$column_index\">";
-	if (sizeof($column_widgets) > 0) {
-		foreach ($column_widgets as $widget) {
-			if (array_key_exists($widget->handler, $widget_types)) {
-				echo elgg_view_entity($widget, array('show_access' => $show_access));
-			}
+for ($column_index = 1; $column_index <= $num_columns; $column_index++) {
+	$column_widgets = (array) elgg_extract($column_index, $widgets, []);
+	
+	$widgets_content = '';
+	foreach ($column_widgets as $widget) {
+		if (!array_key_exists($widget->handler, $widget_types)) {
```

Comment:
thought about this myself too... not doing this for now... `elgg_get_widget_types` has support for exact match, don't want to burden `elgg_get_widgets` with that logic... letting the layout decide this for now... going to do a follow up to remove registering widgets for 'all' context and will remove the exact match then... makes it easier to do a valid handler check in `elgg_get_widgets`